server/src/main/java/org/elasticsearch/repositories/Repository.java

```diff
@@ -306,6 +306,30 @@ void cloneShardSnapshot(
      */
     void awaitIdle();
 
+    /**
+     * Atomically sets the value stored at the given key to {@code updated} if the {@code current value == expected}.
+     * Keys not yet used start at initial value 0.
+     *
+     * @param key key of the value to update
+     * @param expected the expected value
+     * @param updated the new value
+     * @return true if successful, false if the expected value did not match the updated value
+     */
+    default boolean compareAndSetRegister(String key, long expected, long updated) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Gets the value set by {@link #compareAndSetRegister(String, long, long)} for a given key.
+     * If a key has not yet been used, the initial value is 0.
+     *
+     * @param key key of the value to get
+     * @return value found
+     */
+    default long getRegister(String key) throws IOException {
+        throw new UnsupportedOperationException();
+    }
```

Comment:
Should this set and read any byte array rather then a long?

Reply:
In principle a `long` is enough, we can use this as a way to safely create a unique name for some other blob that carries the data we care about. This does add another blob-store round-trip to the process, but I think we are planning to avoid using this heavily (e.g. because of how GCS throttles overwrites).

Reply:
I'm wondering what's the benefit of this layer of indirection, the size of the contents that we'll CAS should be small enough to avoid going through a two steps process, maybe I'm missing something here.

Reply:
I guess one weird downside of allowing any byte-array would be we'd still have to put a limit on its size in some form ? I found that confusing so I liked the fixed length long thing for its simplicity here.
But it's not really a big deal to use a different data type here if the overall plan would benefit from it.

Reply:
Indeed, and in particular the implementation for `FsRepository` here is relying on the write being small enough to be atomic.